### PR TITLE
Support uploading to any S3 service via the WELDR API

### PR DESCRIPTION
**Pull request by ygalblum**
This pull request includes:

- [ ] adequate testing for the new functionality or fixed issue
- [ ] adequate documentation informing people about the change such as
  - [ ] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

### CODE REVIEW

**Comment by ochosi** [comment]
Hi and thanks for your contribution! :)
I know this is still a draft PR but what would really help with reviewing this PR is if you could add some explanation about what it does and why. The same goes for the actual commits (https://www.osbuild.org/guides/developer-guide/workflow.html).
And finally, starting today we will use the PR titles directly for our release notes so ideally the title conveys a clear summary of the changes.

**Comment by ygalblum** [comment]
@ochosi Sorry for that, all the details are in the first commit. I'll paste them here as well and fix the PR's title

**Comment by ygalblum** [comment]
# New Feature
Before this change the worker supported uploading to AWS S3 but only via the REST API. This change enhances this capability in two ways:
1. Support uploading the AWS S3 also via the WELDR API
2. Support non-AWS S3 services (e.g. MinIO)

# Code Changes
## jobimpl-osbuild
Add GenericS3Creds to struct
Add method to create AWS with Endpoint for Generic S3 (with its own credentials file)
Move uploading to S3 and result handling to a separate method
adjust the AWS S3 case to the new method
Implement a new case for uploading to a generic S3 service

## awscloud
Add wrapper methods for endpoint support
Set the endpoint to the AWS session
Set s3ForcePathStyle to true if endpoint was set

## Target
Define a new target type for the GenericS3Target and Options
Handle unmarshaling of the target options and result for the Generic S3

## Weldr
Add support for only uploading to AWS S3
Define new structures for AWS S3 and Generic S3 (based on AWS S3)
Handle unmarshaling of the providers settings' upload settings

## main
Add a section in the main config for the Generic S3 service for credentials
If provided pass the credentials file name to the osbuild job implementation

## Upload Utility
Add upload-generic-s3 utility

## Makefile
Do not fail if the bin directory already exists

**Comment by ygalblum** [comment]
@ochosi Please note that I've set it as a Draft PR in order to be able to run the CI before actually setting it as Ready. But, from any other aspect, it is ready for review.

File: cmd/osbuild-worker/jobimpl-osbuild.go
Comment (thozza) [inline cmd/osbuild-worker/jobimpl-osbuild.go:367]: You don't have to use `options.AWSS3TargetOptions.Bucket`, but can access the `Bucket` from the nested structure directly  as `options.Bucket`. Ditto for the rest of the arguments from the nested `AWSS3TargetOptions` structure.

**Review by thozza — COMMENTED** [review]
Thanks a lot for your contribution, it's a great work 🥳 

I added a few comments inline, but here are some additional comments:

- You forgot to extend `targetsToUploadResponses()` in `internal/weldr/upload.go` for the newly added targets (AWS S3 and generic S3).
- Since you technically enabled also AWS S3 for Weldr API, it would be great if you could add an integration test which would test it. Basically a very similar one to the Generic S3 test case, which you added in this PR, but for AWS S3...

**Comment by thozza** [comment]
I would also like to mention, that this PR will have to reflect #2474 if it is merged after it. Or the other way around... 🙂

File: internal/weldr/upload.go
Comment (ygalblum) [reply internal/weldr/upload.go]: Fixed

File: internal/weldr/upload.go
Comment (ygalblum) [reply internal/weldr/upload.go]: Fixed

File: cmd/osbuild-worker/jobimpl-osbuild.go
Comment (ygalblum) [reply cmd/osbuild-worker/jobimpl-osbuild.go]: Thanks. I'm kinda new with Go. So any comments about the language are welcomed. Done

**Comment by ygalblum** [comment]
>     * You forgot to extend `targetsToUploadResponses()` in `internal/weldr/upload.go` for the newly added targets (AWS S3 and generic S3).
Done

File: cmd/osbuild-worker/jobimpl-osbuild.go
Comment (thozza) [reply cmd/osbuild-worker/jobimpl-osbuild.go]: np 🙂  Thanks!

**Comment by ygalblum** [comment]
> Since you technically enabled also AWS S3 for Weldr API, it would be great if you could add an integration test which would test it. Basically a very similar one to the Generic S3 test case, which you added in this PR, but for AWS S3...

I've implemented the test for AWS. The first run failed and I think I fixed it now. Once the tests pass, I'll move the PR to Ready

**Comment by thozza** [comment]
The Generic S3 test failed - https://gitlab.com/osbuild/ci/osbuild-composer/-/pipelines/500259589

**Comment by ygalblum** [comment]
The new integration tests passed successfully. I've removed the commit that disables all other tests and moved the PR to Ready

**Comment by thozza** [comment]
@ygalblum Thanks a lot!

Unfortunately, with the addition of the aws and generic S2 integration test, we are running into a GitLab CI limitation `jobs:integration:parallel:matrix config generates too many jobs (maximum is 50)`. We'll have to split the integration tests into more groups. We'll discuss this within our team later today...

File: test/cases/aws_s3.sh
Comment (thozza) [inline test/cases/aws_s3.sh:72]: The AWS S3 test should clean up the actual S3 object uploaded to S3 after running `s3_test.sh`. This is not needed for the generic S3, as the object is stored locally in the container, but for AWS, the S3 file would be kept in AWS and just cost us 💰

I propose to add something like https://github.com/osbuild/osbuild-composer/blob/68639b4bf9d679f0ebc180828a1110234913fd03/test/cases/api.sh#L204-L222 to this script's `cleanup()` function.

**Review by thozza — COMMENTED** [review]
I added one comment regarding cleaning up the uploaded object for AWS S3. Otherwise this looks great!

File: test/cases/aws_s3.sh
Comment (ygalblum) [reply test/cases/aws_s3.sh:72]: You're absolutely right. I don't know why, but reading all the shell scripts made me think that there was a garbage collector cleaning up the bucket

Comment (ygalblum) [reply test/cases/aws_s3.sh]: Done

Comment (thozza) [reply test/cases/aws_s3.sh]: We have a scheduled garbage collector running in CI and also one for running after each pipeline. Nevertheless, I think that it is a good practice for the script to clean up after itself. Once can run these test scrips also manually (not that it is user-friendly or easy) and in this case, the object would not get cleaned up.

Thanks a lot for adding the cleanup 😉

**Review by thozza — DISMISSED** [review]
LGTM, although you'll still have to adjust and rebase after #2488 gets merged.

Thanks a lot for your effort!

**Comment by ygalblum** [comment]
> although you'll still have to adjust and rebase after https://github.com/osbuild/osbuild-composer/pull/2488 gets merged
@thozza Done

**Review by thozza — APPROVED** [review]

**Review by thozza — DISMISSED** [review]

**Comment by thozza** [comment]
Rebased using GitHub button, so that the CI infra fixed are picked up...

**Comment by ygalblum** [comment]
It seems like I'm getting some intermittent failures along with a consistent failure in RHEL and CentosStream 9. For some reason, the test fails to connect to the VM and the logs cannot explain why. I'll try to run it locally

**Comment by ygalblum** [comment]
@thozza I finally got my tests to pass. I found that the libvirt tests are defined in a different section of the CI file and since my tests are basically libvirt tests (just with a different storage option), I moved them there. So, all the checks are green and I'm only missing an approval

**Review by thozza — APPROVED** [review]
LGTM. Thanks for all your effort! 🙂

**Comment by thozza** [comment]
@ygalblum would you mind submitting a short documentation update about uploading to S3 in https://github.com/osbuild/guides ?

**Pull request merged**
